ARCs/arc-0060.md

```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
+} => Uint8Array;
 ```
 
-#### Error interface `SignDataError`
+#### Scope `MX_RANDOM`
 
-The `SignDataError` object extends the `SignTxnsError` defined in [ARC-1](./arc-0001.md).
+This scope is inspired by the `signBytes` function in the Algorand SDK which is currently used by some applications. The `signBytes` function allows arbitrary data signing with the only restriction that the data is neither a valid transaction nor a logic signature. This is achieved by enforcing that the first two bytes are the domain separator `MX`. Apart from this restriction, the data is an opaque, unstructured byte string.
+
+Functions `parse`, `validate`, and `serialize` can be defined as follows:
 
 ```ts
-export interface SignDataError extends Error {
-  code: number;
-  data?: any;
-  failingSignData: (StdSigData | null);
+function parse(data: string): Uint8Array {
+    return Buffer.from(data, 'base64');
+}
+function validate(data: Uint8Array): boolean {
+    return data.length >= 2 && data[0] === 0x4D && data[1] === 0x58;
+}
+function serialize(data: Uint8Array): Uint8Array {
+    return data;
 }
 ```
 
-`SignDataError` uses the same error codes of `SignTxnsError` as well as the following codes:
+Since the data is an opaque byte string, wallets cannot inform the user about the actual content that the user is asked to sign. Therefore, it is recommended that applications propose scopes with richer schemas that allow wallets to present more useful information to users.
 
-| Status Code | Name | Description |
-| --- | --- | --- |
-| 4600 | Invalid scope | The scope of the signing action is not recognized by the wallet |
-| 4601 | Invalid schema | The schema does not comply with ARC-60 requirements |
-| 4602 | Invalid encoding | Data cannot be decoded with the given encoding |
+#### EXAMPLE Scope `DOG_OWNERSHIP`
 
-### Semantic Requirements
+This scope is intended purely as an example to showcase how a scope with a stricter schema can be signed. The `DOG_OWNERSHIP` scope signs a JSON object that states that a person owns a dog:
 
-The call `signData(signingData, metadata)` **MUST** either return the signed data `ret` or reject the call throwing an error `err`.
+```ts
+export interface DogOwnership {
+    ownerName: string;
+    dogChipNumber: string;
+}
+```
 
-> Following [ARC-1](./arc-0001.md) terminology, in this ARC the term **Rejecting** means throwing an error with `err.code=4300`.
+We assume that a dog's chip number must be a 15-digit number and that the onwer name must be between 1 and 50 characters long. A `DogOwnership` is serialized according to canonical JSON ([RFC8785](https://www.rfc-editor.org/rfc/rfc8785)), which is intended to seralize JSON deterministically for use in cryptographic operations (hashing, signing, etc). We assume there exists a function `toCanonicalJson` that serializes an object to a byte array according to RFC8785.
 
-Upon calling `signData(signingData,, metadata)`:
+```ts
+function parse(data: string): DogOwnership {
```

Comment:
I don't understand how the schema can "define" these functions though. IMO with a schema, and encoding, we have everything needed to decode and validate a serialized signing object.

What's the need for such functions?

Reply:
> I don't understand how the schema can "define" these functions though.

Somebody puts forth an ARC that defines in plain English or with a reference implementation (preferably both) how the data is parsed, validated, and serialized.

> IMO with a schema, and encoding, we have everything needed to decode and validate a serialized signing object.

Don't you think there are validation rules that cannot be expressed in JSON schema? Like, does the `domain` in an ARC-31 field match the domain of the dApp that the wallet is connected to, etc.

> What's the need for such functions?

As a wallet provider I care about three things: (a) how do I make sense of the data that I get from a dApp, (b) how do I validate if the data is okay according to the specs, and (c) how do I turn the data into bytes such that I can feed them to a signing algorithm. Functions `parse`, `validate`, and `serialize` help me answer those questions.

But perhaps I'm over-complicating things and a JSON schema is sufficient. I don't know. I'd love to finally see a concrete example of structured data (and not just a simple byte array) that needs to be signed in Algorand. Maybe that would help settle this debate